pkg/controller/clone/planner.go

```diff
@@ -0,0 +1,522 @@
+		ds := desiredClaim.Spec.Resources.Requests[corev1.ResourceStorage]
+		is, err := cc.InflateSizeWithOverhead(ctx, p.Client, ds.Value(), &args.TargetClaim.Spec)
+		if err != nil {
+			return nil, err
+		}
+		desiredClaim.Spec.Resources.Requests[corev1.ResourceStorage] = is
+	}
+
+	hcp := &HostClonePhase{
+		Owner:          args.TargetClaim,
+		Namespace:      args.DataSource.Namespace,
+		SourceName:     args.DataSource.Spec.Source.Name,
+		DesiredClaim:   desiredClaim,
+		ImmediateBind:  true,
+		OwnershipLabel: p.OwnershipLabel,
+		Preallocation:  cc.GetPreallocation(ctx, p.Client, args.DataSource.Spec.Preallocation),
+		Client:         p.Client,
+		Log:            args.Log,
+		Recorder:       p.Recorder,
+	}
+
+	if args.DataSource.Spec.PriorityClassName != nil {
+		hcp.PriorityClassName = *args.DataSource.Spec.PriorityClassName
+	}
+
+	rp := &RebindPhase{
+		SourceNamespace: desiredClaim.Namespace,
+		SourceName:      desiredClaim.Name,
+		TargetNamespace: args.TargetClaim.Namespace,
+		TargetName:      args.TargetClaim.Name,
+		Client:          p.Client,
+		Log:             args.Log,
+		Recorder:        p.Recorder,
+	}
+
+	return []Phase{hcp, rp}, nil
+}
+
+func (p *Planner) planSnapshotFromPVC(ctx context.Context, args *PlanArgs) ([]Phase, error) {
+	sourceClaim := &corev1.PersistentVolumeClaim{}
+	exists, err := getResource(ctx, p.Client, args.DataSource.Namespace, args.DataSource.Spec.Source.Name, sourceClaim)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, fmt.Errorf("source claim does not exist")
+	}
+
+	vsc, err := GetCompatibleVolumeSnapshotClass(ctx, p.Client, sourceClaim, args.TargetClaim)
+	if err != nil {
+		return nil, err
+	}
+
+	if vsc == nil {
+		return nil, fmt.Errorf("no compatible volumesnapshotclass")
+	}
+
+	sp := &SnapshotPhase{
+		Owner:               args.TargetClaim,
+		SourceNamespace:     args.DataSource.Namespace,
+		SourceName:          args.DataSource.Spec.Source.Name,
+		TargetName:          fmt.Sprintf("tmp-snapshot-%s", string(args.TargetClaim.UID)),
+		VolumeSnapshotClass: *vsc,
+		OwnershipLabel:      p.OwnershipLabel,
+		Client:              p.Client,
+		Log:                 args.Log,
+	}
+
+	desiredClaim := createDesiredClaim(args.DataSource.Namespace, args.TargetClaim)
+	cfsp := &SnapshotClonePhase{
+		Owner:          args.TargetClaim,
+		Namespace:      args.DataSource.Namespace,
+		SourceName:     sp.TargetName,
+		DesiredClaim:   desiredClaim.DeepCopy(),
+		OwnershipLabel: p.OwnershipLabel,
+		Client:         p.Client,
+		Log:            args.Log,
+		Recorder:       p.Recorder,
+	}
+
+	pcp := &PrepClaimPhase{
+		Owner:           args.TargetClaim,
+		DesiredClaim:    desiredClaim.DeepCopy(),
+		Image:           p.Image,
+		PullPolicy:      p.PullPolicy,
+		InstallerLabels: p.InstallerLabels,
+		OwnershipLabel:  p.OwnershipLabel,
+		Client:          p.Client,
+		Log:             args.Log,
+		Recorder:        p.Recorder,
+	}
+
+	rp := &RebindPhase{
```

Comment:
Maybe it's not necessary, but since we are assembling these phases in a bunch of functions what do you think of adding a constructor function for each phase? In that way we can do something like:

```go
rb := createRebindPhase(source, target, client, log, recorder)
```

I think it can be cleaner and easier to understand.

Reply:
I actually think functions have more opportunity for error especially when there are multiple parameters of the same type.